src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1993,62 +2038,118 @@ SubqueryPushdownMultiPlanTree(Query *queryTree)
-	List *qualifierColumnList = NIL;
-	List *targetListColumnList = NIL;
 	List *columnList = NIL;
 	ListCell *columnCell = NULL;
 	MultiCollect *subqueryCollectNode = CitusMakeNode(MultiCollect);
 	MultiTable *subqueryNode = NULL;
-	MultiSelect *selectNode = NULL;
 	MultiProject *projectNode = NULL;
 	MultiExtendedOp *extendedOpNode = NULL;
 	MultiNode *currentTopNode = NULL;
 	RangeTblEntry *subqueryRangeTableEntry = NULL;
-	List *subqueryEntryList = SubqueryEntryList(queryTree);
-
+	Query *lowerQuery = NULL;
+	List *subqueryTargetEntryList = NIL;
+	List *columnNamesList = NIL;
+	StringInfo rteName = makeStringInfo();
+	List *addedColumnList = NIL;
+	List *havingClauseColumnList = NIL;
 
 	/* verify we can perform distributed planning on this query */
-	ErrorIfQueryNotSupported(queryTree);
-	ErrorIfSubqueryJoin(queryTree);
+	ErrorIfQueryNotSupported(queryTree, true);
 
 	/* extract qualifiers and verify we can plan for them */
 	qualifierList = QualifierList(queryTree->jointree);
 	ValidateClauseList(qualifierList);
 
 	/*
-	 * We disregard pulled subqueries. This changes order of range table list.
-	 * We do not allow subquery joins, so we will have only one range table
-	 * entry in range table list after dropping pulled subquery. For this reason,
-	 * here we are updating columns in the most outer query for where clause
-	 * list and target list accordingly.
+	 * We would be creating a new Query and pushing down top level query's
+	 * contents down to it. Join and filter clauses in higher level query would
+	 * be transferred to lower query. Therefore after this function we would
+	 * only have a single range table entry in the top level query. We need to
+	 * create a target list entry in lower query for each column reference in
+	 * upper level query's target list and having clauses. Any column reference
+	 * in the upper query will be updated to have varno=1, and varattno=<resno>
+	 * of matching target entry in lower query.
 	 */
-	Assert(list_length(subqueryEntryList) == 1);
-
-	qualifierColumnList = pull_var_clause_default((Node *) qualifierList);
-	targetListColumnList = pull_var_clause_default((Node *) targetEntryList);
-
-	columnList = list_concat(qualifierColumnList, targetListColumnList);
+	columnList = pull_var_clause_default((Node *) targetEntryList);
+	havingClauseColumnList = pull_var_clause_default(queryTree->havingQual);
+	columnList = list_concat(columnList, havingClauseColumnList);
 	foreach(columnCell, columnList)
 	{
+		ListCell *addedColumnCell = NULL;
+		StringInfo columnNameString = makeStringInfo();
 		Var *column = (Var *) lfirst(columnCell);
+		bool needsNewTargetEntry = true;
+		AttrNumber resNo = 1;
+
+		/*
+		 * Search if we previously processed this column. We would create a new
+		 * target entry when we encounter the column for the first time. resNo
+		 * is set to resno of the matching target entry for this column in
+		 * lower level query.
+		 */
+		foreach(addedColumnCell, addedColumnList)
```

Comment:
I think it get complicated since I was trying to 2 things at once. I needed to create a target entry for each unique column at the same time updating varattno fields of all columns. Perhaps I should make that in multiple passes.